src/test/java/htsjdk/samtools/util/IOUtilTest.java

```diff
@@ -520,4 +527,240 @@ public void testIsBlockCompressedOnJimfs(Path file, boolean checkExtension, bool
              Assert.assertEquals(IOUtil.isBlockCompressed(jimfsFile, checkExtension), expected);
          }
     }
+
+    @DataProvider
+    public static Object[][] filesToCompress() {
+        return new Object[][]{
+                {TEST_DATA_DIR.resolve("words_longs.txt"), ".gz", 8},
+                {TEST_DATA_DIR.resolve("words_longs.txt"), ".bfq", 8},
+                {TEST_VARIANT_DIR.resolve("test1.vcf"), ".gz", 7},
+                {TEST_VARIANT_DIR.resolve("test1.vcf"), ".bfq", 7}
+
+
+        };
+
+    }
+
+    @Test(dataProvider = "filesToCompress")
+    public void testCompressionLevel(final Path file, final String extension, final int lastDifference) throws IOException {
+        final long origSize = Files.size(file);
+        long previousSize = origSize;
+        for (int compressionLevel = 1; compressionLevel <= 9; compressionLevel++) {
+            final Path outFile = Files.createTempFile("tmp", extension);
+            outFile.toFile().deleteOnExit();
+            IOUtil.setCompressionLevel(compressionLevel);
+            Assert.assertEquals(IOUtil.getCompressionLevel(), compressionLevel);
+            final InputStream inStream = IOUtil.openFileForReading(file);
+
+            try (final OutputStream outStream = IOUtil.openFileForWriting(outFile.toFile())) {
+
+
+                IOUtil.transferByStream(inStream, outStream, origSize);
+            }
+
+
+            final long newSize = Files.size(outFile);
+            if (compressionLevel <= lastDifference) {
+                Assert.assertTrue(previousSize > newSize);
+            } else {
+                Assert.assertTrue(previousSize >= newSize);
+            }
+            previousSize = newSize;
+
+        }
+
+    }
+
+    @DataProvider
+    public static Object[][] badCompressionLevels() {
+        return new Object[][]{
+                {-1},
+                {10}
+        };
+
+    }
+
+    @Test(dataProvider = "badCompressionLevels", expectedExceptions = {IllegalArgumentException.class})
+    public void testCompressionLevelExceptions(final int compressionLevel) {
+        IOUtil.setCompressionLevel(compressionLevel);
+    }
+
+    @DataProvider
+    public static Object[][] filesToCopy() {
+        return new Object[][]{
+                {TEST_VARIANT_DIR.resolve("test1.vcf")},
+                {TEST_DATA_DIR.resolve("ipsum.txt")}
+
+        };
+    }
+
+    @Test(dataProvider = "filesToCopy")
+    public void testCopyFile(final Path file) throws IOException {
+        final Path outFile = Files.createTempFile("tmp", ".tmp");
+        outFile.toFile().deleteOnExit();
+        IOUtil.copyFile(file.toFile(), outFile.toFile());
+
+
+        final Stream<String> stream = Files.lines(file);

```

Comment:
given that we do not need to save on memory here, 
you can replace the remaining lines of code here with a
```java
Assert.assertEquals(File.lines(file),Files.lines(outFile))
```